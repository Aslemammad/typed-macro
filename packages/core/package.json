{
  "name": "@typed-macro/core",
  "version": "1.0.0-alpha",
  "description": "",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/typed-macro/typed-macro"
  },
  "author": "unbyte <i@shangyes.net>",
  "homepage": "https://github.com/typed-macro/typed-macro/tree/master/packages/core#readme",
  "bugs": {
    "url": "https://github.com/typed-macro/typed-macro/issues/new"
  },
  "files": [
    "dist"
  ],
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "dependencies": {
    "@babel/generator": "^7.15.8",
    "@babel/parser": "^7.15.8",
    "@babel/template": "^7.15.4",
    "@babel/traverse": "^7.15.4",
    "@babel/types": "^7.15.6",
    "@types/babel__generator": "^7.6.3",
    "@types/babel__template": "^7.4.1",
    "@types/babel__traverse": "^7.14.2",
    "@typed-macro/shared": "1.0.0-alpha"
  }
}
