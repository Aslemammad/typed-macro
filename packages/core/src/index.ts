export * from './env'
export * from './macro'
export * from './exports'
export * from './provider'
export * from '../types'
