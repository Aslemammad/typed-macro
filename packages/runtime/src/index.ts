export * from './transformer'
export * from './normalizer'
export * from './renderer'
export * from './runtime'
export * from './loader'
export * from './filter'
export * from './hook'
