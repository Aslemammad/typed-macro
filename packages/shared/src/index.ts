export * from './utils'
export * from './version'
export * from './types'
