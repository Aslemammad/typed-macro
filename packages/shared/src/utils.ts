import { existsSync } from 'fs'
import { posix, resolve, sep, win32 } from 'path'
import { PackageManager } from './types'

export function validateFnName(name: string) {
  return /^[_$a-zA-Z\xA0-\uFFFF][_$a-zA-Z0-9\xA0-\uFFFF]*$/.test(name)
}

export function normalizePath(path: string) {
  return path.split(win32.sep).join(posix.sep)
}

export function isString(v: unknown): v is string {
  return typeof v === 'string'
}

export function isPromise(v: unknown): v is Promise<unknown> {
  return Promise.resolve(v) === v
}

export function isError(v: unknown): v is Error {
  return v instanceof Error
}

export function isArray(v: unknown): v is Array<any> {
  return Array.isArray(v)
}

export function isFunction(o: unknown): o is Function {
  return o instanceof Function
}

export function isAsyncFunction(fn: Function) {
  return fn.constructor.name.includes('Async')
}

export function isGeneratorFunction(fn: Function): fn is GeneratorFunction {
  return fn.constructor.name.includes('Generator')
}

/**
 * Get project paths.
 * Directory containing `./package.json` will be considered projects.
 *
 * If current path is under a subproject (for example, in a workspace),
 * the result will be like
 * `[currentProjectDirectory, parentProjectDirectory, ..., rootProjectDirectory]`.
 *
 * If no project directory is found, returns [cwd].
 */
export function getProjectPath() {
  const cwd = process.cwd()
  const paths = cwd.split(sep)
  const result = []
  for (let i = paths.length; i > 0; i--) {
    const path = paths.slice(0, i).join(sep)
    if (existsSync([path, 'package.json'].join(sep))) result.push(path)
  }
  /* istanbul ignore next */
  return result.length ? result : [cwd]
}

/**
 * Get used package manager according to the unique lock file of each package manager.
 *
 * - yarn.lock => yarn
 * - package-lock.json => npm
 * - pnpm-lock.yaml => pnpm
 * - _ => unknown
 */
/* istanbul ignore next */
export function getPackageManager(dir: string): PackageManager {
  if (existsSync(resolve(dir, 'yarn.lock'))) return 'yarn'
  if (existsSync(resolve(dir, 'package-lock.json'))) return 'npm'
  if (existsSync(resolve(dir, 'pnpm-lock.yaml'))) return 'pnpm'
  return 'unknown'
}

/* istanbul ignore next */
// eslint-disable-next-line
export const NO_OP = () => {}
