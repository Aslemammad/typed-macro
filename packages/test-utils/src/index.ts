export * from './transformer'
export * from './runtime'
export * from './renderer'
export * from './env'
