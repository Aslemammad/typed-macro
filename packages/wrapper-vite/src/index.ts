export * from './plugin'
export * from '@typed-macro/core'
