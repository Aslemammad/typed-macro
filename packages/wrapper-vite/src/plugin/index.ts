export { createMacroPlugin, MacroPluginOptions, MacroPlugin } from './plugin'
