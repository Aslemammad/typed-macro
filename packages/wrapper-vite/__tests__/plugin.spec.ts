import { defineMacro, defineMacroProvider } from '@typed-macro/core'
import { createMacroPlugin } from '../src'
import { createMacroRemoveOnProd } from '../../core/__tests__/macros'
import { withTempPath } from '../../runtime/__tests__/utils'

describe('VitePluginMacro', () => {
  const provider = defineMacroProvider({
    id: 'test',
    exports: { '@macros': { macros: [createMacroRemoveOnProd()] } },
  })

  const code = `
    import { dev } from '@macros'
    let a = 0
    dev(a)
    `

  const config = {
    isProduction: true,
    build: {
      ssr: 'some.ts',
    },
  } as any

  it('should work', async () => {
    await withTempPath('./macros.d.ts', async (typesPath) => {
      // dev=false, ssr=true
      {
        const plugin = createMacroPlugin({
          typesPath,
        })

        const onStop = jest.fn()
        // use provider

        plugin.use(
          defineMacroProvider({
            id: 'test',
            exports: { '@macros': { macros: [createMacroRemoveOnProd()] } },
            hooks: { onStop },
          })
        )

        await plugin.configResolved!(config)

        expect(await plugin.resolveId!.call({} as any, '@macros', '', {})).toBe(
          '@macros'
        )
        expect(await plugin.load!.call({} as any, '@macros')).toBe('export {}')

        expect(
          await plugin.transform!.call({} as any, code, 'test.ts')
        ).toMatchSnapshot()

        expect(
          await plugin.transform!.call({} as any, code, 'test.noop')
        ).toBeUndefined()

        await plugin.buildEnd!.call({} as any)
        expect(onStop.mock.calls.length).toBe(1)
      }

      // dev=true, ssr=false
      {
        const plugin = createMacroPlugin({
          typesPath,
          dev: true,
          ssr: false,
        })

        // use provider
        plugin.use(provider)
        await plugin.configResolved!(config)
        expect(
          await plugin.transform!.call({} as any, code, 'test.ts')
        ).toMatchSnapshot()
      }
    })
  })

  it('should support modules/watcher with dev server', async () => {
    // enable server-related
    await withTempPath('./macros.d.ts', async (typesPath) => {
      const plugin = createMacroPlugin({ typesPath })
      let hasModule = false
      let hasWatcher = false
      // use provider
      plugin.use(
        defineMacroProvider({
          id: 'test',
          exports: {
            '@macros': {
              macros: [
                defineMacro('dev')
                  .withSignature('(): void')
                  .withHandler(({ modules, watcher, path }) => {
                    hasModule = !!modules
                    hasWatcher = !!watcher
                    path.remove()
                  }),
              ],
            },
          },
        })
      )

      await plugin.configResolved!(config)
      await plugin.configureServer!({ watcher: {} } as any)
      await plugin.transform!.call({} as any, code, 'test.ts')
      expect(hasWatcher).toBe(true)
      expect(hasModule).toBe(true)
    })
  })
})
