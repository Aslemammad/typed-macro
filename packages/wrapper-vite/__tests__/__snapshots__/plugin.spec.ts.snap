// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VitePluginMacro should work 1`] = `
Object {
  "code": "

let a = 0;",
  "map": null,
}
`;

exports[`VitePluginMacro should work 2`] = `
Object {
  "code": "

let a = 0;
a;",
  "map": null,
}
`;
